prysmaticlabs/prysm
# Simpler and Safer Attester Slashing Protection

**What type of PR is this?**

> Feature

**What does this PR do? Why is it needed?**

Our current slashing protection for attesters suffers from significant lock contention and is not thread-safe as cache invalidation happens at the same time as some data might be accessed from the map itself. We have written up a design doc on what the problem is here: https://hackmd.io/1A2Pb4ZaTuGq2jiNyX3HIw as well as a proposed solution. We decided the best path forward given the time constraint is to have better separation of concerns for saving and retrieving attester history data. We believe this logic should live at the database layer. With this approach, attester protection will work as follows:

### Separation of Concerns: Move the Caching Logic to the DB

In the validator/client package
1. A new slot arrives based on the ticker
2. Fetch duties at that slot (no-op if start of epoch or duties exist)
3. Launch N goroutines for the N assigned duties that you have at that slot
	- 100 attesters, 1 proposer, 30 aggregators
4. During each call to SubmitAttestation(pubKey [48]byte)
	- lock by public key using mputil.NewMultilock()
	- get attestation data from the beacon node
	- fetch the attester history from the DB for that public key
	- check if att is slashable, return err if so
	- we update the history in the DB
	- sign the attestation
	- submit to the beacon node

In the validator/db/kv package
1. In DB NewKVStore() fill a cache of attesting histories by all our public keys we have in our database
3. AttestingHistoryForPubkey()
	- Look into the cache if we have a history then return it
	- If we have a miss, then we look up from the DB and save to cache
4. SaveAttestingHistoryForPubkey()
	- Saves to the cache and the DB

#### Caveats

We will now be using 2Gb of memory for 1000 validating keys in the validator client. Only a few people will be affected by this on mainnet and in Pyrmont. It is a fair tradeoff to make while we work on optimizing slasher.

## Notes

The db cache for attesting history can be disabled with the validator feature flag:

```
--disable-attesting-history-db-cache
```

which we can use to observe just how bad the behavior is without the cache at runtime in the Pyrmont testnet.
